src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc

```diff
@@ -23,153 +23,190 @@ namespace ray {
-    RAY_LOG(DEBUG) << "Node " << node_id.ToInt() << " is "
-                   << (is_available ? "available" : "not available") << " for request "
-                   << resource_request.DebugString()
-                   << " with critical resource utilization "
-                   << critical_resource_utilization << " based on local view "
-                   << node.GetLocalView().DebugString();
-    if (critical_resource_utilization < spread_threshold) {
-      critical_resource_utilization = 0;
-    }
+  return node_resources.IsFeasible(resource_request);
+}
 
-    bool update_best_node = false;
+namespace {
+float ComputeNodeScore(const NodeResources &node_resources, float spread_threshold) {
+  float critical_resource_utilization =
+      node_resources.CalculateCriticalResourceUtilization();
+  if (critical_resource_utilization < spread_threshold) {
+    critical_resource_utilization = 0;
+  }
+  return critical_resource_utilization;
+}
+}  // namespace
 
-    if (is_available) {
-      // Always prioritize available nodes over nodes where the task must be queued first.
-      if (!best_is_available) {
-        update_best_node = true;
-      } else if (critical_resource_utilization < best_utilization_score) {
-        // Break ties between available nodes by their critical resource utilization.
-        update_best_node = true;
-      }
-    } else if (!best_is_available &&
-               critical_resource_utilization < best_utilization_score &&
-               !require_node_available) {
-      // Pick the best feasible node by critical resource utilization.
-      update_best_node = true;
+float HybridSchedulingPolicy::CompuateLocalNodeScore(float spread_threshold) const {
+  const auto local_it = nodes_.find(local_node_id_);
+  RAY_CHECK(local_it != nodes_.end());
+  return ComputeNodeScore(local_it->second.GetLocalView(), spread_threshold);
+}
+
+scheduling::NodeID HybridSchedulingPolicy::GetBestNode(
+    std::vector<std::pair<scheduling::NodeID, float>> &node_scores,
+    size_t num_candidate_nodes,
+    bool prioritize_local_node,
+    float local_node_score) const {
+  RAY_CHECK(!node_scores.empty());
+  RAY_CHECK(num_candidate_nodes >= 1);
+  // Pick the top num_candidate_nodes nodes with the lowest score.
+  // First, sort nodes by id so that we always break ties between nodes in the same
+  // order.
+  std::sort(
+      node_scores.begin(),
+      node_scores.end(),
+      [](const std::pair<scheduling::NodeID, float> &a,
+         const std::pair<scheduling::NodeID, float> &b) { return a.first < b.first; });
+  // Then sort nodes by scores and preserve the order of equivalent elements.
+  std::stable_sort(
+      node_scores.begin(),
+      node_scores.end(),
+      [](const std::pair<scheduling::NodeID, float> &a,
+         const std::pair<scheduling::NodeID, float> &b) { return a.second < b.second; });
+
+  if (prioritize_local_node) {
+    if (local_node_score <= node_scores.front().second) {
```

Comment:
This behavior seems different from before? I think previously the local node was picked even if it had a higher score.

Reply:
previously:
1. if local_node_score < spread threshold, always prefer the local_node.
2. if local_node_score > spread threshold, prefer the node with min score.
So i think the behavior is retained here.